core/src/main/scala/kafka/consumer/BaseConsumer.scala

```diff
@@ -45,21 +57,34 @@ case class BaseConsumerRecord(topic: String,
                               key: Array[Byte],
                               value: Array[Byte])
 
-class NewShinyConsumer(topic: Option[String], whitelist: Option[String], consumerProps: Properties, val timeoutMs: Long = Long.MaxValue) extends BaseConsumer {
+class NewShinyConsumer(topic: Option[String], partitionId: Option[Int], offset: Option[Long], whitelist: Option[String], consumerProps: Properties, val timeoutMs: Long = Long.MaxValue) extends BaseConsumer {
   import org.apache.kafka.clients.consumer.KafkaConsumer
 
   import scala.collection.JavaConversions._
 
   val consumer = new KafkaConsumer[Array[Byte], Array[Byte]](consumerProps)
-  if (topic.isDefined)
-    consumer.subscribe(List(topic.get))
-  else if (whitelist.isDefined)
-    consumer.subscribe(Pattern.compile(whitelist.get), new NoOpConsumerRebalanceListener())
-  else
-    throw new IllegalArgumentException("Exactly one of topic or whitelist has to be provided.")
-
+  consumerInit()
   var recordIter = consumer.poll(0).iterator
 
+  def consumerInit() {
+    if (topic.isDefined)
+      if (partitionId.isDefined) {
+        val topicPartition = new TopicPartition(topic.get, partitionId.get)
+        consumer.assign(List(topicPartition))
+        offset.get match {
+          case OffsetRequest.EarliestTime => consumer.seekToBeginning(List(topicPartition))
+          case OffsetRequest.LatestTime => consumer.seekToEnd(List(topicPartition))
+          case _ => consumer.seek(topicPartition, offset.get)
+        }
+      }
+      else
+        consumer.subscribe(List(topic.get))
+    else if (whitelist.isDefined)
+      consumer.subscribe(Pattern.compile(whitelist.get), new NoOpConsumerRebalanceListener())
+    else
+      throw new IllegalArgumentException("Exactly one of topic or whitelist has to be provided.")
+  }
```

Comment:
@vahidhashemian, I'm late to the party here, but idiomatic Scala would be written like:

```scala
  def consumerInit() {
    (topic, partitionId, offset, whitelist) match {
      case (Some(topic), Some(partitionId), Some(offset), None) =>
        val topicPartition = new TopicPartition(topic, partitionId)
        consumer.assign(List(topicPartition))
        offset match {
          case OffsetRequest.EarliestTime => consumer.seekToBeginning(List(topicPartition))
          case OffsetRequest.LatestTime => consumer.seekToEnd(List(topicPartition))
          case _ => consumer.seek(topicPartition, offset)
        }
      case (Some(topic), None, None, None) =>
        consumer.subscribe(List(topic))
      case (None, None, None, Some(whitelist)) =>
        consumer.subscribe(Pattern.compile(whitelist), new NoOpConsumerRebalanceListener())
      case _ => throw new IllegalArgumentException(...)
    }
  }
```

A couple of notes: the above no longer uses the unsafe `Option.get` method and it also fixes an issue where if we get a `None` for `offset` we get an unhelpful message. Also, the `IllegalArgumentException` needs to be tweaked as we can hit the case if we have a `topic` and `partition`, but no `offset` (you could also have an additional clause for this case separately for an improved error message).

Would you mind doing a `MINOR` follow-up PR including the `toLowerCase` fixes?